R/remove_withdrawn_events.R

```diff
@@ -0,0 +1,110 @@
+#'
+#' @export
+#'
+
+remove_withdrawn_events <- function(dataset, withdrawal_event, withdrawal_date_field, scheduled_event_data, early_discontinuation=FALSE, early_discontinuation_field=NULL, early_discontinuation_value=NULL){
+  
+  #calculate the upper limit of days after the anchor date
+  scheduled_event_data$days_total <- scheduled_event_data$days_after + scheduled_event_data$days_upper
+
+  #group the data based on record id
+  grouped_large <- dataset %>%
+    group_by(record_id)
+  
+  #create a data frame to store all combinations of record_id and event_name
+  event_names <- unique(scheduled_event_data$event_name)
+  combinations <- grouped_large %>%
+    summarize(record_id = first(record_id)) %>%
+    crossing(redcap_event_name = event_names)
+  
+  filtered_data <- dataset
+  
+  #merge necessary columns from event_data with combinations based on redcap_event_name
+  combinations <- merge(combinations, scheduled_event_data, by.x = "redcap_event_name", by.y="event_name", all.x=TRUE )
+  names(filtered_data)[which(names(filtered_data)=="redcap_event_name")] <- "anchor_event"
+  combinations <- merge(combinations, filtered_data, by=c("record_id", "anchor_event"), all.x=TRUE)
+  
+  #function to replace the 'anchor_date' column with the corresponding date value
+  replace_with_date <- function(row) {
+    anchor_date_col <- as.character(row[['anchor_date']])
+    if (!is.na(anchor_date_col) && anchor_date_col %in% names(row)) {
+      row[['anchor_date']] <- row[[anchor_date_col]]
+    }
+    return(row)
+  }
+  
+  #apply the function to each row in the combinations dataset
+  combinations <- t(apply(combinations, 1, replace_with_date))
+  combinations <- as.data.frame(combinations)
+  
+  #select necessary columns from the combinations dataset
+  combinations <- select(combinations, record_id, anchor_event, redcap_event_name, anchor_date, event_date, days_total)
+  
+  #calculate expected_event_date
+  combinations$anchor_date <- as.Date(combinations$anchor_date, format = "%Y-%m-%d")
+  combinations$days_total <- as.numeric(combinations$days_total)
+  combinations$expected_event_date <- combinations$anchor_date + combinations$days_total
+  
+  #merge the expected_event_date with the dataset
+  dataset <- merge(dataset, combinations, by=c("record_id", "redcap_event_name"), all.x=TRUE)
+  
+  #subset dataset to isolate withdsrawal events
+  withdrawal_data <- subset(dataset, redcap_event_name == withdrawal_event)
+  
+  #remove any withdrawal events that are early discontinuations
+  if(early_discontinuation){
+    withdrawal_data[[early_discontinuation_field]] <- as.character(withdrawal_data[[early_discontinuation_field]])
+    withdrawal_data[[early_discontinuation_field]] <- trimws(withdrawal_data[[early_discontinuation_field]])
+    
+    withdrawal_data <- subset(
+      withdrawal_data,
+      withdrawal_data[[early_discontinuation_field]] == early_discontinuation_value
```

Comment:
Doesn't this select values that ARE early discontinuations? Isn't it supposed to remove those?

Reply:
it doesn't, the nomenclature is just bad. I need to update this anyway as the standard withdrawal form has been updated to handle lost to follow up.